### ISSUE TITLE
Allow documentation structure like Haddock does.

### ISSUE THREAD

**Issue by robertmassaioli** [purescript/purescript#2525]
It would be great if the purescript inbuilt documentation format supported documentation structure just like haddock does: https://www.haskell.org/haddock/doc/html/ch03s04.html 

Would that be an idea worth implementing?

**Comment by paf31** [comment]
What does this add over using Markdown sections?

**Comment by robertmassaioli** [comment]
I have the following code:

    -- | The datatype that represents an infinite list. The two type arguments are:
    -- |
    -- | * `a`: The type that will be returned by the getter methods.
    -- | * `b`: The base type for this infinite list that is being iterated over.
    data InfiniteList b a = IL (b -> a) (b -> b) b
    
    -- | ### Generators
    
    -- | Generate an infinite list by providing:
    -- |
    -- |  * a function to iterate from the previous value in the infinite list to the next AND
    -- |  * the first element in the infinite list to begin the iteration.
    -- |
    -- | For example, to generate the list of natural numbers you can do the following:
    -- |
    -- |     naturals = iterate ((+) 1) 1
    -- |
    -- | In english, "The natural numbers start at 1 and you can get the next natural
    -- | number by adding one to the prior number".
    -- |
    -- | Running time: `O(1)`
    iterate :: forall a. (a -> a) -> a -> InfiniteList a a
    iterate iter start = IL id iter start

I expect the generated docs to have the heading for "Generators" to appear before the iterate documentation. Instead this is the generated markdown:

    ``` purescript
    data InfiniteList b a
    ```
    
    The datatype that represents an infinite list. The two type arguments are:
    
    * `a`: The type that will be returned by the getter methods.
    * `b`: The base type for this infinite list that is being iterated over.
    
    ##### Instances
    ``` purescript
    Functor (InfiniteList a)
    Filterable (InfiniteList a)
    ```
    
    #### `iterate`
    
    ``` purescript
    iterate :: forall a. (a -> a) -> a -> InfiniteList a a
    ```
    
    ### Generators
    Generate an infinite list by providing:
    
     * a function to iterate from the previous value in the infinite list to the next AND
     * the first element in the infinite list to begin the iteration.
    
    For example, to generate the list of natural numbers you can do the following:
    
        naturals = iterate ((+) 1) 1
    
    In english, "The natural numbers start at 1 and you can get the next natural
    number by adding one to the prior number".
    
    Running time: `O(1)`

As you can see, the "Generators" heading is considered to be inside the "iterate" function. Which is not what I want. Is there a way to get the "### Generators" heading to appear outside of the iterate function? That is what this issue is about: getting headings outside of specific purescript function documentation.

**Comment by paf31** [comment]
Yes, thanks. I think we would need to parse the Markdown in the comments in order to do this properly. Right now we just treat it like plain text essentially, at least in `psc-docs`. Pursuit uses `cheapskate` to parse and render the content.

We could keep any inline content with the next declaration, but move section headers up.

**Comment by hdgarrood** [comment]
I think I'd prefer to copy what Haddock does, actually, which is that you define sections and headings by putting them in your export list. See the [haddock user guide](https://www.haskell.org/haddock/doc/html/ch03s04.html).

I can't remember if we currently ensure that declarations appear in the same order in the output as they do in the export list (if there is one), but if we don't we'd need to start doing that too.

**Comment by paf31** [comment]
> I can't remember if we currently ensure that declarations appear in the same order in the output as they do in the export list

Does Haskell do that? I hope we wouldn't need to be so strict about export order.

What would the syntax in the exports section look like?

**Comment by hdgarrood** [comment]
Yeah, Haddock does ensure that declarations appear in the same order in the output as they do in the export list. I think this is a good thing to do anyway, but particularly important if we're going to support headings, because if we rearrange things, the headings might not make sense afterwards.

I think we should use the same syntax as Haddock does, except probably switch out `*` for `#` because people are more likely to be used to Markdown, so I think it would look like this:

```purescript
module InfiniteList
  ( InfiniteList
    -- # Generators
  , iterate
  ) where
```

That comment looks like it might be a `<h1>` but I guess we would demote it to whatever was appropriate (probably `<h2>` or `<h3>`) while rendering.

**Comment by paf31** [comment]
Ah, I thought you meant we'd need to make sure the exports matched the order of the declarations in the source. I agree the output should match the exports.

**Comment by matthewleon** [comment]
Are people generally open to pursuing this? I was just today looking at the docs for Haskell's `vector` and thinking to myself how much more accessible the sectioned export list was:

https://hackage.haskell.org/package/vector-0.12.0.1/docs/Data-Vector.html

**Comment by natefaubion** [comment]
Should this be in "Approved"?

**Comment by hdgarrood** [comment]
Yeah, I am definitely in favour of implementing the behaviour I described in https://github.com/purescript/purescript/issues/2525#issuecomment-271170553.

**Comment by hdgarrood** [comment]
Actually I think I'm going to put this into Intend to implement.

**Comment by timjs** [comment]
What about the use case where you'd like to add some explanatory text on a group of declarations?

Say we have `Data.List`, with the functions `elem`, `lookup`, `find`, `filter`, `partition`. They are all about searching lists, the first two with equality, the last three by predicate. With Harry's proposal we will be able add headings, [like in Haddock](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-List.html#g:13), to make the documentation more structured:
```purescript
module Data.List
  ( ..
  -- # Searching lists
  -- ## Searching by equality
  , elem
  , lookup
  -- ## Searching with a predicate
  , find
  , filter
  , partition
  , ..
  )
```

Sometimes it is really convenient to add some explanation about the declarations _in_ a section. I mean something like:
```purescript
module Data.List
  ( ..
  -- # Searching lists
  -- The functions in this section search for elements in a `List`.
  -- All are _O(n)_.
  -- 
  -- ## Searching by equality
  , ..
  )
```

You can make some very clear documentation pages in this way! As an example, take a look at the documentation page of my [`Collage.Layout`](https://package.elm-lang.org/packages/timjs/elm-collage/latest/Collage-Layout) module written for Elm. Every section has an introduction on the concepts used by it's declarations. Elm implements this in a [bit hacky](https://package.elm-lang.org/help/documentation-format) way IMHO. It duplicates exported names in a documentation comment.

An alternative would be to introduce the `export` keyword, and have documentation for each set of exported declarations. We can reuse documentation comments in the way we're already used to, and it gives you a clear overview of the module contents and usage in code too!
```purescript
-- | This module defines a type of strict linked lists,
-- | and associated helper functions and type class instances.
module Data.List

-- | # Searching lists
-- |
-- | The functions in this section search for elements in a `List`.
-- | All are _O(n)_.
-- | 
-- | ## Searching by equality
export elem, lookup

-- | ## Searching with a predicate
export find, filter, partition
```

**Comment by hdgarrood** [comment]
Adding a new keyword is not very appealing to me, but allowing arbitrary markdown within the export list for the purpose of section introductions seems like a good idea.

**Comment by timjs** [comment]
Sounds good to me. We need to think about the structure of these top level comments. I’m above example I used `—` but maybe we should use `— |` for consistency with other doc comments.
